### ISSUE TITLE
Returning a node buffer from native addon kills realtime performance in electron

### ISSUE THREAD

**Issue by justadudewhohacks** [electron/electron#11479]
Me and several users of my native addon are facing performance issues with electron specifically.

More precisely, in the native addon I am creating a node buffer via `Nan::NewBuffer`. Now if I am rebuilding the addon with electron-rebuild `info.GetReturnValue().Set(buf)` seems to be very slow, causing real time processing to be impossible.

For example I am returning a buffer with image data of an 800x320 sized image. In a simple node.js script measured with console.time, returning the node buffer takes less then a millisecond. Doing the exact same thing with electron however takes more than 100ms. The native module is rebuild with electron-rebuild and required via electron.remote: `var cv = require('electron').remote.require('opencv4nodejs')`.

I am assuming that this has to be related to electron as this issue is not apparent with nwjs.

Quick example to reproduce this issue: https://github.com/justadudewhohacks/electron-node-buffer-issue

environment:
- win7 64bit
- node v9.2.0
- electron@1.7.8
- electron-rebuild@1.6.0

**Comment by groundwater** [comment]
Hey @justadudewhohacks what version of nw.js did you use? I wonder if the version of v8/node makes a difference. Node 9.2.0 appears to be running v8 `6.2.414.44` whereas electron 1.7.8 is at `5.8.x`. Can you give this a try on `electron@1.8.2-beta.3`?

**Comment by justadudewhohacks** [comment]
Hey @groundwater, thanks for the quick reply. I just tested with node 9.2 and electron@1.8.2-beta.3 and the issue remains. I didn't try it out personally, but users of my native package have reported that it's working with nw.js, thus they had to change from electron to nw.js.

I edited the issue and appended a link to a repo to reproduce this issue. It seems that the issue does not lie in Nan::NewBuffer but in returning it from the function call.

I will also try it out with nw.js myself and come back at you on this one.

**Comment by groundwater** [comment]
Also just realized you're' calling the `remote` module.

> `var cv = require('electron').remote.require('opencv4nodejs')` 

I think that's going to serialize your array and send it via IPC to the render process. You're better off just requiring it directly in the render process.

**Comment by justadudewhohacks** [comment]
Indeed you are right! I guess the serialization was the issue.

Now in the example repo I am requiring the module simply by: `var m = require('./')`. This solves the performance issue, but now I am facing another issue after calling the module function several times. The application crashes and I am getting the following stacktrace:

``` bash
[8764:1219/215240.897:FATAL:partition_alloc.cc(934)] Check failed: page->num_allocated_slots != -1.
Backtrace:
        GetHandleVerifier [0x000000013FA66D75+23093]
        GetHandleVerifier [0x000000013FAD7803+484547]
        (No symbol) [0x000000013FA5E9C4]
        IsSandboxedProcess [0x0000000142AD20DD+47312277]
        IsSandboxedProcess [0x0000000141A092D7+29712271]
        v8::internal::OptimizingCompileDispatcher::Unblock [0x000007FED8511A21+1825]
        v8::internal::OptimizingCompileDispatcher::Unblock [0x000007FED8511D42+2626]
        v8::internal::AllocationSpaceName [0x000007FED84B7D63+13475]
        v8::internal::AllocationSpaceName [0x000007FED84BFB16+45654]
        v8::internal::Builtins::WasmStackGuard [0x000007FED84AE567+56759]
        v8::internal::AllocationSpaceName [0x000007FED84B823A+14714]
        v8::internal::AllocationSpaceName [0x000007FED84B74EB+11307]
        v8::internal::AllocationSpaceName [0x000007FED84B697E+8382]
        v8::internal::OptimizingCompileDispatcher::IsQueueAvailable [0x000007FED81C0BA9+2841]
        v8::internal::ScavengeJob::NotifyIdleTask [0x000007FED81C2092+1794]
        v8::internal::GCTracer::AddScopeSample [0x000007FED81B4400+13968]
        v8::internal::GCTracer::AddScopeSample [0x000007FED81B40DA+13162]
        v8::internal::OptimizingCompileDispatcher::Enabled [0x000007FED81BBF60+3456]
        v8::internal::interpreter::Bytecodes::ToString [0x000007FED84EE1EC+2652]
        GetHandleVerifier [0x000000013FB05ADD+673693]
        IsSandboxedProcess [0x000000014204F721+36291545]
        IsSandboxedProcess [0x000000014204E59B+36287059]
        IsSandboxedProcess [0x000000014204C01D+36277461]
        GetHandleVerifier [0x000000013FB05ADD+673693]
        GetHandleVerifier [0x000000013FAA6CF6+285110]
        GetHandleVerifier [0x000000013FAA57E0+279712]
        GetHandleVerifier [0x000000013FB07A13+681683]
        GetHandleVerifier [0x000000013FAA2D4E+268814]
        IsSandboxedProcess [0x0000000140149473+3760427]
        GetHandleVerifier [0x000000013FDB22BD+3477373]
        GetHandleVerifier [0x000000013FDB21B4+3477108]
        GetHandleVerifier [0x000000013FB5AA9C+1021788]
        (No symbol) [0x000000013F7B7E8F]
        IsSandboxedProcess [0x0000000142A2C13F+46632439]
        BaseThreadInitThunk [0x00000000774459CD+13]
        RtlUserThreadStart [0x000000007767A561+33]
```

I tried with electron 1.7.0 and 1.8.2-beta.3 and it appears on both versions. When I comment out the call to Nan::NewBuffer in the example repo, this crash does seem to not happen.

I am running chrome v63.0.

**Comment by starpit** [comment]
FYI we're also seeing the num_allocated_slots failure on 1.8.2-beta.3. We haven't seen this with the 1.6 or 1.7 series. It happens when a DOM element vertically overflows, and becomes a scroll region.

**Comment by justadudewhohacks** [comment]
Hey some news on this problem,

Turns out that the issue might lie in freeing the memory held by the buffer. Apparently creating a buffer with: `Nan::NewBuffer(data, size)` will use a default callback, which frees the memory. Using this constructor will almost immediately crash the app. However, explicitly providing a callback as follows seems to work fine:
``` c++
void callback(char* data, void* hint) {
  free(data);
}

... 
Nan::NewBuffer(data, size, callback, 0)
```
As this issue only occurs with electron however, I am not quite sure if it's related to electron or an issue with the nan buffer itself.

**Comment by MarshallOfSound** [comment]
I think this is due to how memory is freed in a blink environment, seem to recall @zcbenz talking about this at some point 🤔

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v2.0.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons.

Thanks in advance! Your help is appreciated.

**Comment by sofianguy** [comment]
Thank you for your issue!

We haven't gotten a response to our questions in our comment above. With only the information that is currently in the issue, we don't have enough information to take action. I'm going to close this but don't hesitate to reach out if you have or find the answers we need, we'll be happy to reopen the issue.
